motion_primitives_forward_controller/src/motion_primitives_forward_controller.cpp

```diff
@@ -0,0 +1,554 @@
+// Copyright (c) 2025, b»robotized
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Authors: Mathias Fuhrer
+
+#include "motion_primitives_forward_controller/motion_primitives_forward_controller.hpp"
+
+#include <limits>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "controller_interface/helpers.hpp"
+#include "motion_primitives_forward_controller/execution_state.hpp"
+#include "motion_primitives_forward_controller/motion_type.hpp"
+#include "motion_primitives_forward_controller/ready_for_new_primitive.hpp"
+
+namespace
+{  // utility
+
+// TODO(destogl): remove this when merged upstream
+// Changed services history QoS to keep all so we don't lose any client service calls
+static constexpr rmw_qos_profile_t rmw_qos_profile_services_hist_keep_all = {
+  RMW_QOS_POLICY_HISTORY_KEEP_ALL,
+  1,  // message queue depth
+  RMW_QOS_POLICY_RELIABILITY_RELIABLE,
+  RMW_QOS_POLICY_DURABILITY_VOLATILE,
+  RMW_QOS_DEADLINE_DEFAULT,
+  RMW_QOS_LIFESPAN_DEFAULT,
+  RMW_QOS_POLICY_LIVELINESS_SYSTEM_DEFAULT,
+  RMW_QOS_LIVELINESS_LEASE_DURATION_DEFAULT,
+  false};
+}  // namespace
+
+namespace motion_primitives_forward_controller
+{
+MotionPrimitivesForwardController::MotionPrimitivesForwardController()
+: controller_interface::ControllerInterface()
+{
+}
+
+controller_interface::CallbackReturn MotionPrimitivesForwardController::on_init()
+{
+  RCLCPP_INFO(get_node()->get_logger(), "Initializing Motion Primitives Forward Controller");
```

Comment:
Bear in mind that this controller will be part of a potentially large set of controllers to be started. There's a lot of output going on already and the controller manager will also say which controller it is loading. So maybe make that a debug output?